paddle/fluid/framework/tensor_util.cc

```diff
@@ -495,6 +495,127 @@ void TensorFromStream(std::istream& is, Tensor* tensor,
+                 : (type.code == kDLUInt)
+                       ? static_cast<void*>(
+                             dst->mutable_data<uint8_t>(dst_place))
+                       : nullptr;
+    case 16:
+      return (type.code == kDLInt)
+                 ? static_cast<void*>(dst->mutable_data<int16_t>(dst_place))
+                 : (type.code == kDLFloat)
+                       ? static_cast<void*>(
+                             dst->mutable_data<paddle::platform::float16>(
+                                 dst_place))
+                       : nullptr;
+    case 32:
+      return (type.code == kDLInt)
+                 ? static_cast<void*>(dst->mutable_data<int32_t>(dst_place))
+                 : (type.code == kDLFloat)
+                       ? static_cast<void*>(dst->mutable_data<float>(dst_place))
+                       : nullptr;
+    case 64:
+      return (type.code == kDLInt)
+                 ? static_cast<void*>(dst->mutable_data<int64_t>(dst_place))
+                 : (type.code == kDLFloat)
+                       ? static_cast<void*>(
+                             dst->mutable_data<double>(dst_place))
+                       : nullptr;
+    default:
+      break;
+  }
+
+  return nullptr;
+}
+
+#ifdef PADDLE_WITH_CUDA
+void* GetDstPtrByDLDataType(DLDataType type, framework::Tensor* dst,
+                            const platform::CUDAPlace& dst_place) {
+  if (type.lanes > 1) return nullptr;  // vector types not currently supported
+
+  switch (type.bits) {
+    case 8:
+      return (type.code == kDLInt)
+                 ? static_cast<void*>(dst->mutable_data<int8_t>(dst_place))
+                 : (type.code == kDLUInt)
+                       ? static_cast<void*>(
+                             dst->mutable_data<uint8_t>(dst_place))
+                       : nullptr;
+    case 16:
+      return (type.code == kDLInt)
+                 ? static_cast<void*>(dst->mutable_data<int16_t>(dst_place))
+                 : (type.code == kDLFloat)
+                       ? static_cast<void*>(
+                             dst->mutable_data<paddle::platform::float16>(
+                                 dst_place))
+                       : nullptr;
+    case 32:
+      return (type.code == kDLInt)
+                 ? static_cast<void*>(dst->mutable_data<int32_t>(dst_place))
+                 : (type.code == kDLFloat)
+                       ? static_cast<void*>(dst->mutable_data<float>(dst_place))
+                       : nullptr;
+    case 64:
+      return (type.code == kDLInt)
+                 ? static_cast<void*>(dst->mutable_data<int64_t>(dst_place))
+                 : (type.code == kDLFloat)
+                       ? static_cast<void*>(
+                             dst->mutable_data<double>(dst_place))
+                       : nullptr;
+    default:
+      break;
+  }
+
+  return nullptr;
+}
+#endif
+
+void TensorFromDLPack(const ::DLTensor& dl_tensor,
+                      const platform::DeviceContext& ctx,
```

Comment:
Seems that you can use `ctx` to get `CUDADeviceContext::stream()` inside the function. You can try to use:

`platform::DeviceContextPool::Instance().GetByPlace(platform::CUDAPlace(device_id))`

to get `CUDADeviceContext`. So that, you do not need to parse `const platform::DeviceContext &ctx` in this function.

The reason why I prefer not to parse  `const platform::DeviceContext &ctx` is that both `dl_tensor` and `ctx` carry the device information. They may be conflict with each other if the device id is not the same.

Reply:
Good Idea!
Done!